### PR TITLE
Lp 1927656

### PR DESCRIPTION
**Pull request by ycliuhw**
*Ensure the controllerUUID annotation of the migrated model's namespace gets updated to the new controller's UUID;*

## Checklist

 - [ ] ~Requires a [pylibjuju](https://github.com/juju/python-libjuju) change~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed~
 - [x] Comments answer the question of why design decisions were made

## QA steps

```console
$ juju bootstrap microk8s k1

$ juju bootstrap microk8s k2

$ juju add-model t1 -c k1 && juju deploy cs:~juju/mariadb-k8s-3 -m k1:t1

$ yml2json ~/.local/share/juju/controllers.yaml | jq '.controllers|to_entries|map("\(.key) => \(.value.uuid)")'
[
  "k2 => 16e28e79-9ccd-471f-8bd8-2ed35e53bb14",
  "k1 => e9f2ceb6-4d1e-44db-8c7d-5c5aaf34fe1f"
]

$ mkubectl get ns t1 -o json | jq .metadata.annotations
{
  "controller.juju.is/id": "e9f2ceb6-4d1e-44db-8c7d-5c5aaf34fe1f",
  "model.juju.is/id": "c688bb92-0587-4ad0-8209-431e9e5a65e7"
}

$ juju migrate k1:t1 k2

$ mkubectl get ns t1 -o json | jq .metadata.annotations
{
  "controller.juju.is/id": "16e28e79-9ccd-471f-8bd8-2ed35e53bb14",
  "model.juju.is/id": "c688bb92-0587-4ad0-8209-431e9e5a65e7"
}

$ juju kill-controller -t 0 -y k2 --debug

$ mkubectl get ns t1
Error from server (NotFound): namespaces "t1" not found

```

## Documentation changes

No

## Bug reference

https://bugs.launchpad.net/juju/+bug/1927656

### CODE REVIEW

**Comment by ycliuhw** [comment]
[migrated CaaS model will left in the cluster after model destroyed](https://trello.com/c/E9Vadxsp/1643-migrated-caas-model-will-left-in-the-cluster-after-model-destroyed)

File: caas/kubernetes/provider/base_test.go [hunk 1]
```diff
@@ -202,14 +201,14 @@ func (s *BaseSuite) setupController(c *gc.C) *gomock.Controller {
 		return "appuuid", nil
 	}
 
-	s.mockNamespaces.EXPECT().Get(gomock.Any(), s.getNamespace(), v1.GetOptions{}).
+	s.mockNamespaces.EXPECT().Get(gomock.Any(), s.getNamespace(), v1.GetOptions{}).Times(2).
```
Comment (hpidcock) [inline caas/kubernetes/provider/base_test.go:204]: I think just making this .Any() would be fine.

File: caas/kubernetes/provider/k8s.go [hunk 2]
```diff
@@ -212,13 +212,48 @@ func newK8sBroker(
...
+		return errors.NewNotValid(nil,
+			fmt.Sprintf("annotations %v for namespace %q must include %v", ns.GetAnnotations(), k.namespace, k.annotations),
+		)
+	}
+	annotationControllerUUIDKey := utils.AnnotationControllerUUIDKey(isLegacy)
+	k.annotations.Add(annotationControllerUUIDKey, controllerUUID)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if ns.GetAnnotations()[annotationControllerUUIDKey] == controllerUUID {
```
Comment (hpidcock) [inline caas/kubernetes/provider/k8s.go:242]: There's a few references to `ns.GetAnnotations()`, maybe store the value in a variable and refer to that instead.

**Review by hpidcock — APPROVED** [review]
LGTM, couple of non-critical comments

File: caas/kubernetes/provider/base_test.go (hunk 1 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/base_test.go:204]: I'd prefer asserting more explicitly. And also set to AnyTimes() here might fail the later namespace get assertion statements in other real tests.

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**